### PR TITLE
Add edit command

### LINKED ISSUE #18: As a user I should be able to mark a book as read
so that I can keep track of what to read and what I have read.

### LINKED ISSUE #20: As a user I should be able to change the priority of a book
so that I can keep track of what to read next.

### LINKED ISSUE #25: As a user I should be able to rate books
so that I can give recommendations to others.

### PR DESCRIPTION
**Pull request by 592363789**
* Add edit command to edit the status, priority, and rating of books
* Add unit and integration tests for edit command

Fixes #18 
Fixes #20 
Fixes #25

### CODE REVIEW

File: src/main/resources/view/BookDetailsPanel.fxml [hunk 1]
```diff
@@ -44,6 +44,7 @@
           <Label styleClass="label-muted" text="Categories" GridPane.rowIndex="2"/>
           <Label styleClass="label-muted" text="Publisher" GridPane.rowIndex="3"/>
           <Label styleClass="label-muted" text="Publication Date" GridPane.rowIndex="4"/>
+          <Label styleClass="label-muted" text="Rating" GridPane.rowIndex="5"/>
```
Comment (qiu-siqi) [inline src/main/resources/view/BookDetailsPanel.fxml]: remove.

File: src/test/java/seedu/address/logic/parser/EditCommandParserTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        expectedCommand = new EditCommand(INDEX_FIRST_BOOK, new Rating(EMPTY_RATING),
+                Priority.LOW, Status.UNREAD);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_missingCompulsoryField_failure() throws Exception {
+        String expectedMessage = String.format(MESSAGE_INVALID_COMMAND_FORMAT, EditCommand.MESSAGE_USAGE);
+
+        // no parameters
```
Comment (qiu-siqi) [inline src/test/java/seedu/address/logic/parser/EditCommandParserTest.java]: add another test with `EditCommand.COMMAND_WORD + " 1"`

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 3]
```diff
@@ -56,7 +55,7 @@ private void handleKeyPress(KeyEvent keyEvent) {
             navigateToNextInput();
             break;
         default:
-            // let JavaFx handle the keypress
+                // let JavaFx handle the keypress
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/ui/CommandBox.java]: indent back?

File: src/main/java/seedu/address/ui/BookDetailsPanel.java [hunk 4]
```diff
@@ -66,6 +66,7 @@ private void showBook(Book book) {
             isbn.setText(book.getIsbn().toString());
             publisher.setText(book.getPublisher().toString());
             publicationDate.setText(book.getPublicationDate().toString());
+            rating.setText((book.getRating().toString()));
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/ui/BookDetailsPanel.java]: remove (it's already below).

File: src/main/java/seedu/address/ui/BookCard.java [hunk 5]
```diff
@@ -29,21 +29,22 @@
...
     private Label priority;
-    @FXML
-    private Label rating;
 
     public BookCard(Book book, int displayedIndex) {
         super(FXML);
         this.book = book;
         id.setText(displayedIndex + ". ");
         title.setText(book.getTitle().title);
+        rating.setText(String.valueOf(book.getRating().rating));
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/ui/BookCard.java]: remove (it's already below).

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        this.index = index;
+        this.rating = rating;
+        this.priority = priority;
+        this.status = status;
+    }
+
+    @Override
+    public CommandResult executeUndoableCommand() throws CommandException {
+        requireAllNonNull(bookToEdit, editedBook);
+        if (model.getActiveListType() != ActiveListType.BOOK_SHELF) {
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: this should go in `preprocessUndoableCommand`, before checking for valid index.

File: docs/DeveloperGuide.adoc [hunk 7]
```diff
@@ -458,6 +458,29 @@ image::UndoRedoActivityDiagram.png[width="550"]
 **Additional Info:** See the discussion  https://github.com/se-edu/addressbook-level4/issues/390#issuecomment-298936672[here].
 // end::undoredo[]
 
+=== Rate for recent books: `rate` command
+
+==== Current implementation
+
+When `rate` command is entered, eventually `SwitchToRateBooksRequestEvent` is posted to the
```
Comment (qiu-siqi) [inline docs/DeveloperGuide.adoc]: `SwitchToRateBooksRequestEvent` doesn't exist? o.o

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+import seedu.address.model.ActiveListType;
+import seedu.address.model.book.Book;
+import seedu.address.model.book.Priority;
+import seedu.address.model.book.Rating;
+import seedu.address.model.book.Status;
+import seedu.address.model.book.exceptions.BookNotFoundException;
+import seedu.address.model.book.exceptions.DuplicateBookException;
+
+/**
+ * Changes the rating of an existing Book in the address book.
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: `Edits the status, priority, and rating of an existing book.`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ */
+public class EditCommand extends UndoableCommand {
+
+    public static final String COMMAND_WORD = "edit";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the rating of the book identified "
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Change the description

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            + "by the index number used in the last book listing. "
+            + "Existing rating will be overwritten by the input.\n"
+            + "Parameters: INDEX (must be a positive integer) "
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: `Parameters: [s/STATUS] [p/PRIORITY] [r/RATING]`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_RATING + "-1" + PREFIX_PRIORITY + "low" + PREFIX_STATUS
+            + "unread";
+
+    public static final String MESSAGE_ADD_EDITING_SUCCESS = "Editing to Book: %1$s";
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Just use a single message `MESSAGE_EDIT_SUCCESS = "Edited book: %1$s."`

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    private final Index index;
+    private final Rating rating;
+    private final Priority priority;
+    private final Status status;
+    private Book bookToEdit;
+    private Book editedBook;
+
+    /**
+     * @param index of the book in the filtered book list to edit the rating
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Update this line. Add `@param` description for `priority` and `status`. Make sure they all end with a full-stop.

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+                status, priority, rating,
+                bookToEdit.getPublisher(),
+                bookToEdit.getPublicationDate());
+    }
+
+    /**
+     * Generates a command execution success message based on whether the rating is added to or removed from
+     * {@code bookToEdit}.
+     */
+    private String generateSuccessMessage(Book bookToEdit) {
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/commands/EditCommand.java]: Remove this method.

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import seedu.address.commons.core.index.Index;
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.logic.commands.EditCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.book.Priority;
+import seedu.address.model.book.Rating;
+import seedu.address.model.book.Status;
+
+/**
+ * Parses input arguments and creates a new {@code EditCommand} object
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: Add a full-stop.

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public class EditCommandParser implements Parser<EditCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the {@code EditCommand}
+     * and returns a {@code RateCommand} object for execution.
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: Add a full-stop.

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+import seedu.address.model.book.Rating;
+import seedu.address.model.book.Status;
+
+/**
+ * Parses input arguments and creates a new {@code EditCommand} object
+ */
+public class EditCommandParser implements Parser<EditCommand> {
+    /**
+     * Parses the given {@code String} of arguments in the context of the {@code EditCommand}
+     * and returns a {@code RateCommand} object for execution.
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: `EditCommand`

File: src/main/java/seedu/address/logic/parser/EditCommandParser.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
+        } catch (IllegalValueException ive) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, EditCommand.MESSAGE_USAGE));
+        }
+
+        int rating = Integer.valueOf(argMultimap.getValue(PREFIX_RATING).orElse("-1"));
+        String priority = argMultimap.getValue(PREFIX_PRIORITY).orElse("n");
+        String status = argMultimap.getValue(PREFIX_STATUS).orElse("u");
+        return new EditCommand(index, new Rating(rating), Priority.findPriority(priority),
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/EditCommandParser.java]: Check if the priority, status, and rating entered by the user is valid.
If the user input is invalid, throw a `ParseException`. You can refer to `ListCommandParser`.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 18]
```diff
@@ -13,9 +13,10 @@
 import seedu.address.model.book.Author;
 import seedu.address.model.book.Category;
 import seedu.address.model.book.Description;
-import seedu.address.model.book.Rating;
 import seedu.address.model.book.Title;
 
+
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: Remove these 2 lines.

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 19]
```diff
@@ -14,7 +14,6 @@
 import seedu.address.logic.commands.CommandResult;
 import seedu.address.logic.commands.exceptions.CommandException;
 import seedu.address.logic.parser.exceptions.ParseException;
-
```
Comment (takuyakanbr) [inline src/main/java/seedu/address/ui/CommandBox.java]: Add back this line.

File: src/test/java/guitests/guihandles/BookCardHandle.java [hunk 20]
```diff
@@ -17,11 +17,11 @@
 public class BookCardHandle extends NodeHandle<Node> {
     private static final String ID_FIELD_ID = "#id";
     private static final String TITLE_FIELD_ID = "#title";
+    private static final String RATING_FIELD_ID = "#rating";
```
Comment (takuyakanbr) [inline src/test/java/guitests/guihandles/BookCardHandle.java]: Move this line below `PRIORITY_FIELD_ID`.

File: src/test/java/seedu/address/logic/commands/EditCommandTest.java [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.book.Priority;
+import seedu.address.model.book.Rating;
+import seedu.address.model.book.Status;
+import seedu.address.network.NetworkManager;
+
+/**
+ * Contains integration tests (interaction with the Model) and unit tests for editCommand.
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/EditCommandTest.java]: `EditCommand`

File: src/test/java/seedu/address/logic/commands/EditCommandTest.java [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+ */
+public class EditCommandTest {
+
+    public static final int RATING_STUB = -1;
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/EditCommandTest.java]: Remove this.

File: src/test/java/seedu/address/logic/commands/EditCommandTest.java [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        assertFalse(standardCommand.equals(new EditCommand(INDEX_SECOND_BOOK, new Rating(VALID_RATING_ARTEMIS),
+                Priority.DEFAULT_PRIORITY, Status.DEFAULT_STATUS)));
+
+        // different Rating -> returns false
+        assertFalse(standardCommand.equals(new EditCommand(INDEX_FIRST_BOOK, new Rating(VALID_RATING_BABYLON),
+                Priority.DEFAULT_PRIORITY, Status.DEFAULT_STATUS)));
+    }
+
+    /**
+     * Returns an {@code editCommand} with parameters {@code index} and {@code Rating}.
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/commands/EditCommandTest.java]: `Returns an {@code EditCommand} with the specified parameters.`

File: src/test/java/seedu/address/logic/parser/BookShelfParserTest.java [hunk 24]
```diff
@@ -37,7 +44,16 @@ public void parseCommand_delete() throws Exception {
         DeleteCommand command = (DeleteCommand) parser.parseCommand(DeleteCommand.COMMAND_WORD + " 1");
         assertEquals(new DeleteCommand(INDEX_FIRST_BOOK), command);
     }
-
+    @Test
+    public void parseCommand_edit() throws Exception {
+        final Rating rating = new Rating(-1);
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/parser/BookShelfParserTest.java]: Why are these 3 variables needed?

File: src/test/java/seedu/address/logic/parser/EditCommandParserTest.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+import seedu.address.model.book.Rating;
+import seedu.address.model.book.Status;
+
+public class EditCommandParserTest {
+    private static final int EMPTY_RATING = -1;
+    private static final int NON_EMPTY_RATING = 5;
+    private EditCommandParser parser = new EditCommandParser();
+
+    @Test
+
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/parser/EditCommandParserTest.java]: Remove this empty line.

File: src/test/java/seedu/address/logic/parser/EditCommandParserTest.java [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+        expectedCommand = new EditCommand(INDEX_FIRST_BOOK, new Rating(EMPTY_RATING),
+                Priority.LOW, Status.UNREAD);
+        assertParseSuccess(parser, userInput, expectedCommand);
+    }
+
+    @Test
+    public void parse_missingCompulsoryField_failure() throws Exception {
+        String expectedMessage = String.format(MESSAGE_INVALID_COMMAND_FORMAT, EditCommand.MESSAGE_USAGE);
+
+        //no parameters
```
Comment (takuyakanbr) [inline src/test/java/seedu/address/logic/parser/EditCommandParserTest.java]: Add a space before `no`.

File: docs/DeveloperGuide.adoc [hunk 27]
```diff
@@ -462,6 +462,33 @@ image::UndoRedoActivityDiagram.png[width="550"]
 **Additional Info:** See the discussion  https://github.com/se-edu/addressbook-level4/issues/390#issuecomment-298936672[here].
 // end::undoredo[]
 
+=== edit for recent books: `edit r/rating p/pirority s/status` command
```
Comment (takuyakanbr) [inline docs/DeveloperGuide.adoc]: `Edit status, priority, and rating of books`

File: docs/DeveloperGuide.adoc [hunk 28]
```diff
@@ -462,6 +462,33 @@ image::UndoRedoActivityDiagram.png[width="550"]
...
+The user performs `select 1` to view information on the 1st book in the book shelf.
+User can see the default rating is -1, the default priority is NONE, the default status is UNREAD.
+
+The user performs `rate 1 r/ 1 p/ l s/ u ` to rate 1st with 1 point, low priority and unread.
+
+==== Design considerations
+
+===== Aspect: What to show to user about the rating
+
+* **Alternative 1 (current choice):** Show user the number of rating
```
Comment (takuyakanbr) [inline docs/DeveloperGuide.adoc]: Check that every line ends with a full-stop.

**Pull request merged**